Some fine tuning of paren spacing

jsnajdr opened the pull request:
- for parens that don't line-break, turn the added spaces into separate atoms. In future, they will be added conditionally when an option is enabled -- this change makes it easier.
- for parens that can line-break, play with `softline`/`line` instead of adding spaces -- that leads to more reliable formatting without bugs
- fine-tune the unary expression spacing to make it conform better to how `space-unary-ops` is configured in `eslint-config-wpcalypso`

@samouri Would you mind trying out this branch and checking if the results are better? I'm only aware about one bug:
```js
map( a =>
  a(),
 ); // extra space
```
This case is quite tough, I hope to solve it in the next day or two.

My dream here is to introduce a `calypsoSpacing` option for prettier, and make the fork support both the enabled and disabled state -- with the enabled state being the hardcoded default for us.

That would make a merge into the official version at least theoretically possible.

To test, check that the following rules as defined in `eslint-config-wpcalypso` are always happy:
- `computed-property-spacing`
- `space-in-parens`
- `space-unary-ops`
- `jsx-curly-spacing`
- `template-curly-spacing`
- `object-curly-spacing`

jsnajdr committed:
Some fine tuning of paren spacing

- for parens that don't line-break, turn the added spaces into separate atoms.
  In the next step, they will be added conditionally when an option is enabled.
- for parens that can line-break, play with softline/line -- that leads to more
  reliable formatting without bugs
- tune the unary expression spacing to make it conform better to how space-unary-ops
  is configured in eslint-config-wpcalypso

samouri commented on src/printer.js:
```diff
@@ -908,8 +907,10 @@ function genericPrintNoParens(path, options, print, args) {
-      if ( n.argument && n.argument.type !== "UnaryExpression" && ( n.operator !== '+' && n.operator !== '-') ) {
-        parts.push(' ');
+
+      if (/[a-z]$/.test(n.operator)) parts.push(" ");
```
what is this catching other than `typeof`?
Shouldn't it the conditional immediately below this be an `else if`

jsnajdr replied:
I made this change to match the [configuration](https://github.com/Automattic/eslint-config-wpcalypso/blob/master/index.js#L81) of the `space-unary-ops` eslint rule. It adds space after textual unary ops (`typeof`, `delete`) and `!`, handles the `!!` special case, and doesn't add space after others -- `~`, `+`, `-` etc.

Both `if` conditions here should be unified into one with `||` op.

Before this patch, only the `~` operator and its combinations didn't work correctly, so it's not a big deal.

samouri commented on src/printer.js:
```diff
@@ -1249,10 +1257,10 @@ function genericPrintNoParens(path, options, print, args) {
           typeof n.value.value === "string"
         ) {
-          const value = n.value.extra ? n.value.extra.raw : n.value.value;
+          const value = n.value.extra ? n.value.extra.raw : n.value.raw;
```
why are they both `raw`?

jsnajdr replied:
I just reverted your previous changes here, which were most likely a mistake.

samouri commented:
Thanks for working on this @jsnajdr.

Initial thoughts:
1. is there any benefit of using `line` instead of `" " + softline`?  While line makes more sense for the original prettier project, I don't love the idea of having two different way for adding in extra calypso spacing.
2. I think it would be a really good idea to add in the variable `calypsoSpace` now.  If the flag is true, make it `" "`, else empty string `""`.  That way it will be blatantly obvious on inspection if it were added for calypso or in the original prettier.
3. We'll soon need to start being careful about how often we make changes calypso-prettier. It would be frustrating for any teams using this if they were hit by whitespace changes regularly.

jsnajdr committed:
Regenerated docs

jsnajdr committed:
Correct spacing for expanded last arg

Add a slightly complicated logic that handles correct Calypso-style
spacing for expanded last arg of a function:

```
fun( 1, 2, action => next( action ) ); // calypso-space before closing paren
fun( 1, 2, action =>
  next( action )
); // always newline before closing paren
fun( 1, 2, action => [
  3
] ); // calypso-space before closing paren
```

jsnajdr committed:
Add calypso-spacing command line option

Turn the hardcoded spaces into an option. Makes the changes suitable
for potential merge to official prettier.

jsnajdr commented:
> is there any benefit of using line instead of " " + softline? While line makes more sense for the original prettier project, I don't love the idea of having two different way for adding in extra calypso spacing.

`line` is exactly the same thing as `" " + softline`, so there is no difference in behavior -- `line` is just a very useful shortcut. I think it's a good idea to keep the same conventions as the rest of the prettier codebase.

There are indeed two ways how to add the Calypso space -- one adds a potential line break, the other add just an unbreakable whitespace. And it's intentional -- it's a way to express that we want the outer `()` to 'hug' the inner `{}` or `[]`. For example, there's no break between the nested brackets here:
```
fun( {
  param1,
  param2
} );
```

> I think it would be a really good idea to add in the variable calypsoSpace now. If the flag is true, make it " ", else empty string "". That way it will be blatantly obvious on inspection if it were added for calypso or in the original prettier.

I pushed a commit that adds a `calypso-spacing` option that does exactly this. Our prettier fork has all options hardcoded and ignores all options passed from outside, but if we ever want to honor the options, everything is ready.

> We'll soon need to start being careful about how often we make changes calypso-prettier. It would be frustrating for any teams using this if they were hit by whitespace changes regularly.

Fortunately, the prettier maintainers feel the same way. This is what @vjeux told me on a Gitter chat the other day:

> by the way, now that we reach 1.0, I'm very sensitive around formatting changes and upgrade velocity, so they should start changing less and less things and the releases be more spaced apart.

By the way, what do you think about upgrading our fork to the latest 1.4.x release? When I find a bug or a potential improvement in the 1.1 version that we use, it's almost always fixed in the upstream version.

samouri commented on bin/prettier.js:
```diff
@@ -121,14 +122,15 @@ function getTrailingComma() {
 }
 
 const options = {
-  useTabs: true,
```
will need to rebase after you merge #3

samouri commented on src/doc-builders.js:
```diff
@@ -46,6 +46,7 @@ function group(contents, opts) {
     type: "group",
     contents: contents,
     break: !!opts.shouldBreak,
+    addedLine: !!opts.addedLine,
```
what is this for?

samouri replied:
found the comments at https://github.com/Automattic/calypso-prettier/pull/1/files#diff-76f26752c3761490942405c0704a8e1dR357.  Its definitely a level above my knowledge of how prettier works internally.

I'll do my best to try to understand it but won't slow down 🚢  because of it!
I'll do some testing later today

jsnajdr replied:
Yeah, wrapping the last function argument is one of the most complex tricks in Prettier and took me some time to understand :) You can just test if the formatting doesn't break in some weird cases.

samouri commented on src/printer.js:
```diff
@@ -114,13 +114,13 @@ function genericPrint(path, options, printPath, args) {
 
   if (needsParens) {
-    parts.unshift("( ");
+    parts.unshift("(", options.calypsoSpacing ? " " : "");
```
Wouldn't it be simplest if at the top of the file you had:
`const calypsoSpace = options.calypsoSpacing ? " " : ""`

Then in all cases where you have this ternary logic you could write something like:
`parths.unshift( "(", calypsoSpace );`

samouri replied:
just saw that you basically do this later by creating `parenSpace`.  I'm still in favor of hoisting it one scope + of renaming to calypsoSpace to both line up with `calypsoSpacing` option and to jump out at maintainers of this package

jsnajdr replied:
> Wouldn't it be simplest if at the top of the file you had:
> `const calypsoSpace = options.calypsoSpacing ? " " : ""`

`options` is a parameter to `genericPrint` and `genericPrintNoParens`, so these definitions can't be at top of the file. I'll repeat the definition of `calypsoSpace` at the beginning of both print functions.

jsnajdr replied:
About the naming: wouldn't `wpSpacing` and `wp-prettier` be better names than `calypsoSpacing` and `calypso-prettier`? This spacing style is really a WordPress thing, used also in non-Calypso code like the Gutenberg editor or JS in WordPress core.

samouri commented:
> By the way, what do you think about upgrading our fork to the latest 1.4.x release? When I find a bug or a potential improvement in the 1.1 version that we use, it's almost always fixed in the upstream version.

Upgrades are welcome :). I'm personally a bit checked out of this project because it already solves main gripe pretty well (manual formatting), but I'm happy to review any work you do to make it better.
If you want to update to 1.4 you should make a new branch in this repo called "calypso-1.4"

samouri reviewed (approved):
I tested this on all of calypso client code and results were great 💯. Merge away!
